StructuredConcurrency.md

```diff
@@ -0,0 +1,283 @@
+address this, leading to the proposal
+["Non-cooperative goroutine preemption"](https://github.com/golang/proposal/blob/master/design/24543-non-cooperative-preemption.md).
+In the process, several interesting alternatives
+[were assessed](https://github.com/golang/go/issues/24543) including cooperative
+preemption of loops (by the insertion of safe points) and more complex
+mechanisms such as returning from a signal to out-of-line code which leads
+quicly to a safe point.
+
+## Syntax
+
+When comparing to solutions in other languages it's important to mention that
+many have introduced special syntax to mark concurrent code.
+
+* C# introduced `async`/`await`; many followed (Python, Rust, ...). This makes
+  potential suspension points syntactic.
+* `await` in Python marks preemption points. `async` is required to go with it,
+  forming a chain of custody around "potentially suspending" functions.
+* Kotlin has `suspend` to introduce a special calling convention which passes
+  along the coroutine context.
+* Go doesn't have `async` or `await` but is deeply concurrent and is the best
+  analogy to Julia.
+
+The problem with `async`/`suspend` is that it splits the world of functions in
+two, as nicely expressed in Bob Nystrom's blog post
+["What color is your function?"](http://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/).
+This is a barrier to composability because higher order functions have to know
+about the color of the function they're being passed. Bob argues that Go
+handles this in the nicest way by having first class support for continuations
+in the language. The Julia runtime does this in the same way.
+
+On the other hand, a syntax such as `async`/`await` is arguably a useful visual
+marker for possible cancellation points (`await`) and for which functions are
+cancellable (`async`). Note that this doesn't have to be implemented at the
+language level; for example, Go's context and errgroup also allow the reader to
+recognize where the cancellation can happen (listening to the Done channel) and
+which functions can be cancelled (those that accept Context as an argument).
+
+## Prototypical use cases
+
+* The "happy eyeballs" algorithm is becoming a standard example of structured
+  concurrency thanks to `@njsmith`'s tutorial at PyCon 2018.
+  - [General discussion](https://trio.discourse.group/t/happy-eyeballs-structured-concurrencys-hello-world/57)
+  - [Trio implementation](https://github.com/python-trio/trio/blob/master/trio/_highlevel_open_tcp_stream.py)
+  - [Libdill implementation](https://github.com/sustrik/libdill/blob/master/happyeyeballs.c) and [discussion](http://250bpm.com/blog:139)
+* The Go concurrency tutorial — in his talk `@elizarov` suggested that
+  implementing all the examples there was a great inspiriation.
+
+## Related julia issues and prototypes
+
+* [Tapir parrallel IR](https://github.com/JuliaLang/julia/pull/31086)
```

Comment:
Note that this is primarily about:
- Compiler representation of tasks
- Experimenting with Cilk-style semantics (e.g. may happen in parallel or what I have taken to call structured parallelism)

Reply:
Thanks, I've replied to this in the main repo.